File: src/lib/notify.ts
import { getAdminClient } from "@/services/supabase-admin";
import type { NotificationType } from "@/types/database";

interface NotificationPayload {
  type: NotificationType;
  title: string;
  body?: string;
  data?: Record<string, unknown>;
}

// Insert notifications for a set of profile IDs (deduped). Uses admin client to bypass RLS.
async function insertNotifications(profileIds: string[], payload: NotificationPayload) {
  const unique = [...new Set(profileIds.filter(Boolean))];
  if (unique.length === 0) return;
  const admin = getAdminClient();
  await admin.from("notifications").insert(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    unique.map((profile_id) => ({ profile_id, ...payload })) as any,
  );
}

// Returns profile IDs of all active admins.
async function getAdminProfileIds(): Promise<string[]> {
  const admin = getAdminClient();
  const { data } = await admin.from("profiles").select("id").eq("role", "admin").eq("is_active", true);
  return (data ?? []).map((r) => r.id);
}

// Notify a specific user + all admins (admin deduped if they are the actor).
export async function notifyUserAndAdmins(
  recipientId: string | null,
  payload: NotificationPayload,
  excludeId?: string,
) {
  const adminIds = await getAdminProfileIds();
  const ids = [...(recipientId ? [recipientId] : []), ...adminIds].filter((id) => id !== excludeId);
  await insertNotifications(ids, payload);
}

// Notify only admins.
export async function notifyAdmins(payload: NotificationPayload, excludeId?: string) {
  const adminIds = await getAdminProfileIds();
  await insertNotifications(
    adminIds.filter((id) => id !== excludeId),
    payload,
  );
}

File: backend/app/core/notify.py
from typing import Any, Optional, cast

from app.core.supabase import get_admin_db_client


def notify_user_and_admins(
    recipient_id: Optional[str],
    payload: dict,
    exclude_id: Optional[str] = None,
) -> None:
    admin = get_admin_db_client()
    raw = admin.table("profiles").select("id").eq("role", "admin").eq("is_active", True).execute()
    rows: list[dict[str, Any]] = cast(list[dict[str, Any]], raw.data or [])
    admin_ids = [r["id"] for r in rows]
    ids = list(
        dict.fromkeys(
            i
            for i in ([recipient_id] if recipient_id else []) + admin_ids
            if i and i != exclude_id
        )
    )
    if ids:
        admin.table("notifications").insert(
            [{"profile_id": pid, **payload} for pid in ids]
        ).execute()


def notify_admins(payload: dict, exclude_id: Optional[str] = None) -> None:
    notify_user_and_admins(None, payload, exclude_id)
